Rubberduck.Parsing/ComReflection/ComTypedElement.cs

```diff
@@ -0,0 +1,165 @@
+        public virtual object Value { get; private set; }
+        public virtual bool IsReferenceType { get; private set; }
+        public virtual bool IsArray { get; private set; }
+
+        private string _type = string.Empty;
+
+        public virtual string TypeName
+        {
+            get
+            {
+                var typeName = _type;
+                if (IsEnumMember && ComProject.KnownEnumerations.TryGetValue(EnumGuid, out ComEnumeration enumType))
+                {
+                    var member = enumType.Members.FirstOrDefault(m => m.Value == Value);
+                    if (member != null)
+                    {
+                        typeName = member.TypeName;
+                    }                    
+                }
+                else if (IsAlias && ComProject.KnownAliases.TryGetValue(AliasGuid, out ComAlias alias))
+                {
+                    typeName = alias.Name;
+                }
+                return IsArray ? $"{typeName}()" : typeName;
+            }
+        }
+
+        protected Guid EnumGuid { get; private set; } = Guid.Empty;
+        public virtual bool IsEnumMember => !EnumGuid.Equals(Guid.Empty);
+        protected Guid AliasGuid { get; private set; } = Guid.Empty;
+        public virtual bool IsAlias => !AliasGuid.Equals(Guid.Empty);
+
+        protected ComTypedElement(ITypeInfo info, VARDESC varDesc, string name, int index, bool constant = false)
+        {
+            Name = name;
+            Index = index;
+
+            if (constant)
+            {
+                GetValue(varDesc.desc.lpvarValue);
+            }
+            else
+            {
+                GetFieldType(varDesc.elemdescVar.tdesc, info);
+            }
+        }
+
+        protected ComTypedElement(ITypeInfo info, VARDESC varDesc) 
+        {
+            var names = new string[1];
+            info.GetNames(varDesc.memid, names, names.Length, out int count);
+            Debug.Assert(count == 1);
+            Name = names[0];
+            GetValue(varDesc.desc.lpvarValue);
+        }
+
+        protected ComTypedElement(ITypeInfo info, ELEMDESC elemDesc, string name, int index) : this(info, elemDesc.tdesc)
+        {
+            Name = name;
+            Index = index;
+        }
+
+        protected ComTypedElement(ITypeInfo info, TYPEDESC desc)
+        {
+            GetFieldType(desc, info);
+        }
+
+        private void GetValue(IntPtr lpvarValue)
+        {
+            var value = new ComVariant(lpvarValue);
+            Value = value.Value;
+            _type = value.TypeName;
+            IsArray = value.VariantType.HasFlag(VarEnum.VT_ARRAY);
+        }
+
+        private void GetFieldType(TYPEDESC desc, ITypeInfo info)
+        {
+            // ReSharper disable once SwitchStatementMissingSomeCases  <-- Don't care, the processed cases are the exceptions, default is the rule.
+            switch ((VarEnum)desc.vt)
+            {
+                case VarEnum.VT_PTR:
```

Comment:
Just checking - according to [this doc](https://docs.microsoft.com/en-us/windows/desktop/api/wtypes/ne-wtypes-varenum), `TYPEDESC` doesn't use `VT_PTR`. Does this case get hit? (I'd think it's a documentation error but want to be absolutely sure). I also have same question regarding whether it'll be used as a bitmask.

Reply:
I thought it does get hit, but I'd need to verify - that's shared with the PARAMDESC type resolution, so I'm not 100% sure that there would be another way to signify a parameter that took a pointer.